src/compute/src/server.rs

```diff
@@ -143,18 +203,42 @@ pub fn serve(
                 .send(channels)
                 .expect("worker should not drop first");
         }
-        let client =
-            LocalClient::new_partitioned(response_rxs, command_txs, worker_threads.clone());
-        Box::new(client) as Box<dyn ComputeClient>
-    };
-    let server = Server {
-        _worker_guards: worker_guards,
-    };
-    Ok((server, client_builder))
+        self.inner = Some(LocalClient::new_partitioned(
+            response_rxs,
+            command_txs,
+            worker_threads,
+        ));
+        self.worker_guards = Some(worker_guards);
+        Ok(())
+    }
 }
 
-type CommandReceiver = crossbeam_channel::Receiver<ComputeCommand>;
-type ResponseSender = mpsc::UnboundedSender<ComputeResponse>;
+#[async_trait]
+impl GenericClient<ComputeCommand, ComputeResponse> for ClusterClient<PartitionedClient> {
+    async fn send(&mut self, cmd: ComputeCommand) -> Result<(), Error> {
+        match cmd {
+            cmd @ ComputeCommand::CreateInstance(_) => {
+                self.build()?;
+                self.inner.as_mut().expect("intialized").send(cmd).await
+            }
+            cmd @ ComputeCommand::DropInstance => {
+                self.inner.as_mut().expect("intialized").send(cmd).await?;
+                self.inner = None;
+                let _ = self.worker_guards.take().map(|guards| guards.join());
+                Ok(())
+            }
+            cmd => self.inner.as_mut().expect("intialized").send(cmd).await,
```

Comment:
It is not necessary to re-bind `cmd` in each branch

```suggestion
            ComputeCommand::CreateInstance(_) => {
                self.build()?;
                self.inner.as_mut().expect("intialized").send(cmd).await
            }
            ComputeCommand::DropInstance => {
                self.inner.as_mut().expect("intialized").send(cmd).await?;
                self.inner = None;
                let _ = self.worker_guards.take().map(|guards| guards.join());
                Ok(())
            }
            _ => self.inner.as_mut().expect("intialized").send(cmd).await,
```

Reply:
The rules of what match is a move and what's not are still surprising to me! Thanks, fixed.